packages/ledgerstate/output.go

```diff
@@ -0,0 +1,1031 @@
+		WriteBytes(s.address.Bytes()).
+		Bytes()
+}
+
+// String returns a human readable version of the Output.
+func (s *SigLockedColoredOutput) String() string {
+	return stringify.Struct("SigLockedColoredOutput",
+		stringify.StructField("id", s.ID()),
+		stringify.StructField("address", s.address),
+		stringify.StructField("balances", s.balances),
+	)
+}
+
+// code contract (make sure the type implements all required methods)
+var _ Output = &SigLockedColoredOutput{}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region CachedOutput /////////////////////////////////////////////////////////////////////////////////////////////////
+
+// CachedOutput is a wrapper for the generic CachedObject returned by the objectstorage that overrides the accessor
+// methods with a type-casted one.
+type CachedOutput struct {
+	objectstorage.CachedObject
+}
+
+// Retain marks the CachedObject to still be in use by the program.
+func (c *CachedOutput) Retain() *CachedOutput {
+	return &CachedOutput{c.CachedObject.Retain()}
+}
+
+// Unwrap is the type-casted equivalent of Get. It returns nil if the object does not exist.
+func (c *CachedOutput) Unwrap() Output {
+	untypedObject := c.Get()
+	if untypedObject == nil {
+		return nil
+	}
+
+	typedObject := untypedObject.(Output)
+	if typedObject == nil || typedObject.IsDeleted() {
+		return nil
+	}
+
+	return typedObject
+}
+
+// Consume unwraps the CachedObject and passes a type-casted version to the consumer (if the object is not empty - it
+// exists). It automatically releases the object when the consumer finishes.
+func (c *CachedOutput) Consume(consumer func(output Output), forceRelease ...bool) (consumed bool) {
+	return c.CachedObject.Consume(func(object objectstorage.StorableObject) {
+		consumer(object.(Output))
+	}, forceRelease...)
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region OutputMetadata ///////////////////////////////////////////////////////////////////////////////////////////////
+
+// OutputMetadata contains additional Output information that are derived from the local perception of the node.
+type OutputMetadata struct {
+	id                      OutputID
+	branchID                BranchID
+	branchIDMutex           sync.RWMutex
+	solid                   bool
+	solidMutex              sync.RWMutex
+	solidificationTime      time.Time
+	solidificationTimeMutex sync.RWMutex
+	consumerCount           int
+	firstConsumer           TransactionID
+	consumerMutex           sync.RWMutex
+	preferred               bool
+	preferredMutex          sync.RWMutex
+	liked                   bool
+	likedMutex              sync.RWMutex
+	finalized               bool
+	finalizedMutex          sync.RWMutex
+	confirmed               bool
+	confirmedMutex          sync.RWMutex
+	rejected                bool
+	rejectedMutex           sync.RWMutex
```

Comment:
These are mutually exclusive and could be locked with a single mutex. In that way we avoid weird behaviour like an output being `confirmed` and `rejected` at the same time. 

What I'm saying is this: the model could support us by preventing to get into inconsistent states. I.e. there could be some logic in here that helps with developing and debugging.

Reply:
I would like to keep the models as dumb as possible but would it maybe make sense to introduce an enum for this kind of "final state" (i.e. Pending, Confirmed, Rejected) where it can have exactly one of those values?

`Pending`: `confirmed = false && rejected == false` (in the old model)
`Confirmed`: `confirmed = true && rejected = false` (in the old model)
`Rejected`: `confirmed = false && rejected = true` (in the old model)

Then the invalid state of `confirmed = true && rejected = true` wouldn't even exist

Reply:
According to our discussion in the slack channel, we will need to revisit this once we introduce grades of finality and levels of knowledge anyway and leave it like this for now.